In [25]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)

In [26]:
# Crear un dataframe
df = pd.read_csv('CSV/hr_raw_data_final.csv', index_col=0) ## Añadimos index col para eliminar columna 'Unnamed'

In [28]:
df.head(10)

age attrition     businesstravel    dailyrate                department  \
0  51        No                NaN  2015.722222                       NaN   
1  52        No                NaN  2063.388889                       NaN   
2  42        No      travel_rarely  1984.253968   Research & Development    
3  47        No      travel_rarely  1771.404762                       NaN   
4  46        No                NaN  1582.771346                       NaN   
5  48        No                NaN  1771.920635   Research & Development    
6  59        No                NaN  1032.487286                       NaN   
7  42        No      travel_rarely   556.256661                       NaN   
8  41        No                NaN  1712.182540                       NaN   
9  41        No  travel_frequently  1973.984127                       NaN   

   distancefromhome  education    educationfield  employeecount  \
0                 6          3               NaN              1   
1                 1          4     Life Sciences              1   
2                 4          2  Technical Degree              1   
3                 2          4           Medical              1   
4                 3          3  Technical Degree              1   
5                22          3           Medical              1   
6                25          3     Life Sciences              1   
7                 1          1               NaN              1   
8                 2          5               NaN              1   
9                 9          3               NaN              1   

   employeenumber  environmentsatisfaction  gender  hourlyrate  \
0               1                        1       0         NaN   
1               2                        3       0         NaN   
2               3                        3       0         NaN   
3               4                        1       1         NaN   
4               5                        1       1         NaN   
5               6                        4       1         NaN   
6               7                        1       1         NaN   
7               8                        2       0   69.532083   
8               9                        2       1         NaN   
9              10                        1       0         NaN   

   jobinvolvement  joblevel              jobrole  jobsatisfaction  \
0               3         5   resEArch DIREcToR                 3   
1               2         5             ManAGeR                 3   
2               3         5             ManaGER                 4   
3               3         4   ReseArCH DIrECtOr                 3   
4               4         4     sAleS EXECUtIve                 1   
5               3         4             MANAger                 4   
6               3         3     Sales ExeCutIVe                 1   
7               3         2     Sales eXEcUTiVe                 3   
8               3         4             mANAGEr                 1   
9               3         5   reSEaRCH DIrectoR                 3   

  maritalstatus monthlyincome monthlyrate  numcompaniesworked over18 overtime  \
0           NaN     16280,83$   42330,17$                   7      Y       No   
1           NaN           NaN   43331,17$                   0    NaN      NaN   
2       Married           NaN   41669,33$                   1    NaN       No   
3       Married     14307,50$   37199,50$                   3      Y      NaN   
4      Divorced     12783,92$   33238,20$                   2      Y       No   
5           NaN     14311,67$   37210,33$                   3    NaN       No   
6           NaN      8339,32$   21682,23$                   7      Y      NaN   
7       Married           NaN   11681,39$                   1    NaN       No   
8       Married     13829,17$   35955,83$                   7    NaN       No   
9           NaN     15943,72$   41453,67$                   2    NaN       No   

   percentsalaryhike performanceratin

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1678 entries, 0 to 1677
Data columns (total 41 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   age                       1678 non-null   object 
 1   attrition                 1678 non-null   object 
 2   businesstravel            877 non-null    object 
 3   dailyrate                 1678 non-null   float64
 4   department                312 non-null    object 
 5   distancefromhome          1678 non-null   int64  
 6   education                 1678 non-null   int64  
 7   educationfield            904 non-null    object 
 8   employeecount             1678 non-null   int64  
 9   employeenumber            1678 non-null   int64  
 10  environmentsatisfaction   1678 non-null   int64  
 11  gender                    1678 non-null   int64  
 12  hourlyrate                411 non-null    float64
 13  jobinvolvement            1678 non-null   int64  
 14  joblevel     

In [5]:
# df columnas numéricas
df_numericos = pd.read_csv('CSV/numericos.csv', index_col=0)

In [6]:
df_numericos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1678 entries, 0 to 1677
Data columns (total 29 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   age                       1678 non-null   int64  
 1   dailyrate                 1678 non-null   float64
 2   distancefromhome          1678 non-null   int64  
 3   education                 1678 non-null   int64  
 4   employeecount             1678 non-null   int64  
 5   employeenumber            1678 non-null   int64  
 6   environmentsatisfaction   1678 non-null   int64  
 7   hourlyrate                411 non-null    float64
 8   jobinvolvement            1678 non-null   int64  
 9   joblevel                  1678 non-null   int64  
 10  jobsatisfaction           1678 non-null   int64  
 11  monthlyincome             1189 non-null   float64
 12  monthlyrate               1678 non-null   float64
 13  numcompaniesworked        1678 non-null   int64  
 14  percentsalary

In [10]:
df_numericos

age    dailyrate  distancefromhome  education  employeecount  \
0      51  2015.722222                 6          3              1   
1      52  2063.388889                 1          4              1   
2      42  1984.253968                 4          2              1   
3      47  1771.404762                 2          4              1   
4      46  1582.771346                 3          3              1   
...   ...          ...               ...        ...            ...   
1673   43   488.944444               -26          3              1   
1674   47  1973.984127                26          4              1   
1675   29   290.035510                15          3              1   
1676   47  1032.487286                 4          3              1   
1677   32   556.256661                 2          2              1   

      employeenumber  environmentsatisfaction  hourlyrate  jobinvolvement  \
0                  1                        1         NaN               3   
1                  2                        3         NaN               2   
2                  3                        3         NaN               3   
3                  4                        1         NaN               3   
4                  5                        1         NaN               4   
...              ...                      ...         ...             ...   
1673             824                        2         NaN               4   
1674            1087                        4         NaN               3   
1675             528                        3         NaN               3   
1676              76                        3         NaN               2   
1677             401                        4   69.532083               3   

      joblevel  ...  trainingtimeslastyear  worklifebalance  yearsatcompany  \
0            5  ...                      5              3.0              20   
1            5  ...                      5              3.0              33   
2            5  ...                      3              NaN              22   
3            4  ...                      2              NaN              20   
4            4  ...                      5              3.0              19   
...        ...  ...                    ...              ...             ...   
1673         1  ...                      2              3.0               3   
1674         5  ...                      2              3.0               5   
1675         1  ...                      1              3.0               6   
1676         3  ...                      4              3.0              22   
1677         2  ...                      2              3.0               8   

      yearsincurrentrole  yearssincelastpromotion  yearswithcurrmanager  \
0                    NaN                       15                    15   
1                    NaN                       11                     9   
2                    NaN                       11                    15   
3                    NaN                        5                     6   
4                    NaN                        2                     8   
...                  ...                      ...                   ...   
1673                 NaN                        1                     2   
1674                 NaN                        1                     0   
1675                 NaN                        1                     5   
1676                 NaN                       14                    10   
1677                 NaN                        0                     7   

      sameasmonthlyincome  datebirth     salary  numberchildren  
0                16280.83       1972  195370.00             NaN  
1                     NaN       1971  199990.00             NaN  
2                     NaN       1981  192320.00             NaN  
3                14307.50       1976  171690.00             NaN  
4                12783.92       1977        NaN             

In [23]:
df_numericos.columns

Index(['age', 'dailyrate', 'distancefromhome', 'education', 'employeecount',
       'employeenumber', 'environmentsatisfaction', 'hourlyrate',
       'jobinvolvement', 'joblevel', 'jobsatisfaction', 'monthlyincome',
       'monthlyrate', 'numcompaniesworked', 'percentsalaryhike',
       'performancerating', 'relationshipsatisfaction', 'stockoptionlevel',
       'totalworkingyears', 'trainingtimeslastyear', 'worklifebalance',
       'yearsatcompany', 'yearsincurrentrole', 'yearssincelastpromotion',
       'yearswithcurrmanager', 'sameasmonthlyincome', 'datebirth', 'salary',
       'numberchildren'],
      dtype='object')

## Información columnas y posibles cambios: 

#### 1) Columna 'age': 

Info: 
- Descripción: La edad del empleado
- Datatype: int64
- Duplicados: N/A (puede repetirse por empleado)
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: 
- Columna similar: N/A
- Mantener como columna numérica? Si

#### 2) Columna 'dailyrate': 

Info: 
- Descripción: La tarifa diaria del empleado.
- Datatype: float64 (no redondeado) 
- Duplicados: N/A (el salario diario puede repetirse por empleado)
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: 
- Columna similar: N/A
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
- Mover el decimal una posición a la izquierda (EJ: fila 1677, el hourly rate es 69.532083 y el dailyrate es 556.256661. Si movemos el decimal y dividimos 55.6256661 por 8, da 6.9532083)
- Redondear a dos decimales? 

#### 3) Columna 'distancefromhome': 

Info: 
- Descripción: La distancia desde el hogar del empleado hasta su lugar de trabajo.
- Datatype: int64
- Duplicados: N/A (la distancia puede repetirse por empleado)
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: 
- Columna similar: N/A
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
- Hay distancias negativas (202). Verificar si es posible cambiar a positivo (los km no pueden ser negativos)

-- Contar cuántos valores negativos hay en la columna "distancefromhome": 
negativos = (df["distancefromhome"] < 0).sum()
print(f"Número de valores negativos en 'distancefromhome': {negativos}")-- 

#### 4) Columna 'education': 

Info: 
- Descripción: Nivel de educación del empleado (generalmente en una escala del 1 al 5).
- Datatype: int64
- Duplicados: N/A (la education puede repetirse por empleado)
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: No en esta, pero quizas si en la columna educationfield por medio del education level. 
- Columna similar: educationfield?
- Mantener como columna numérica? Si
- Unique: [3, 4, 2, 1, 5]
- Value counts: 
3    649
4    461
2    322
1    186
5     60

Observaciones / Posibles cambios:
- Ver si tiene relación con la columna categorica 'educationfield'. Ver en comprobaciones el código de contingencia: 

#Ver contingencia entre las dos tablas: (para verificar si hay relacion entre education y field) 
contingency_table = pd.crosstab(df['education'], df['educationfield'])
contingency_table

#### 5) Columna 'employeecount': 

Info: 
- Descripción: Un contador que generalmente es 1 y se usa para contar empleados.
- Datatype: int64
- Duplicados: N/A (siempre es 1)
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A
- Columna similar: N/A
- Mantener como columna numérica? Si
- Unique: 1
- Value counts: N/A

Observaciones / Posibles cambios:
- Ver si es útil o podemos eliminarlo. 

#### 6) Columna 'employeenumber': 

Info: 
- Descripción: Un número de identificación único para el empleado.
- Datatype: int64
- Duplicados: 64
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A
- Columna similar: N/A
- Mantener como columna numérica? Si
- Unique: n/a (son muchos)
- Value_counts: n/a (son muchos)

Observaciones / Posibles cambios:
- Eliminar duplicados. 


#### 7) Columna 'environmentsatisfaction': 

Info: 
- Descripción: Nivel de satisfacción del empleado en relación con su entorno de trabajo. Con valores que estan comprendidos entre el 1 y el 4, siendo el 4 el nivel de máxima satisfacción. 
- Datatype: int64
- Duplicados: 0
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A
- Columna similar: N/A
- Mantener como columna numérica? Si
- Unique: *Deberia ser del 1 al 4, pero va del 1 al 49*
- Value_counts: (son muchos)

Observaciones / Posibles cambios:
- Ver si tiene que ver con el departamento o otra columna para modificar los datos. O si es posible que el 49 sea 4,9 o algo asi. 


#### 8) Columna hourlyrate: 

Info: 
- Descripción: La tarifa por hora del empleado.
- Datatype: float64
- Duplicados: N/A
- Nulos: 1267 / En caso se ser mayoría de nulos, puede completarse con otra columna?: (ver con el resto de las columnas salariales)
- Columna similar: N/A
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
- **************

## Comprobaciones

In [70]:
contingency_table = pd.crosstab(df['education'], df['educationfield'])
contingency_table

educationfield  Human Resources  Life Sciences  Marketing  Medical  Other  \
education                                                                   
1                             0             42          9       36      5   
2                             1             67         15       64     12   
3                             6            133         40      116     21   
4                             4            115         34       64     22   
5                             1             10          8        6      3   

educationfield  Technical Degree  
education                         
1                             10  
2                             17  
3                             27  
4                             14  
5                              2

In [29]:
# Contar cuántos valores negativos hay una columna
negativos1 = (df_numericos["hourlyrate"] < 0).sum()
print(f"Número de valores negativos en 'hourlyrate': {negativos1}")

Número de valores negativos en 'hourlyrate': 0


In [30]:
# Verificar duplicados
df_numericos['hourlyrate'].duplicated().sum()


1483

In [31]:
#Ver nulos
df_numericos['hourlyrate'].isnull().sum()

1267

In [32]:
df_numericos["hourlyrate"].unique()

array([         nan,  69.53208262, 172.84325397, 216.04761905,
        79.97321429, 129.06091077, 246.74801587, 207.17460317,
        60.8859127 ,  36.25443872, 139.99503968,  86.24107143,
       197.84641831,  61.06646825, 245.62003968, 123.97718254,
       148.44246032, 255.96329365, 134.41071429,  71.92559524,
       113.78869048, 175.10019841, 247.49007937, 220.83234127,
        91.34821429,  37.77480159, 198.9593254 , 124.05456349,
       247.58035714, 108.48809524,  43.17857143,  68.46924603,
        32.48710317,  55.26289683,  83.46825397,  79.26388889,
        72.28670635,  45.96428571,  31.66170635,  65.19345238,
        52.24503968, 143.9156746 , 113.40178571, 143.19345238,
       139.86607143,  70.57142857,  97.04861111,  45.02281746,
       223.47619048, 134.9781746 ,  74.40178571,  67.14087302,
        58.06150794,  35.05357143,  73.18948413, 164.33134921,
        43.53968254,  54.79861111, 252.59722222,  83.81646825,
        46.71230159,  33.00297619,  70.77777778,  69.23

In [ ]:
df_numericos['XXXXXXXXXX'].value_counts()